File: src/simulation/FormationOptimizer.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Simulation;

/// <summary>
/// Greedy optimizer that assigns the best-fit player to each formation slot
/// based on position-specific overall ratings.
/// </summary>
public static class FormationOptimizer
{
    /// <summary>
    /// Returns 11 player IDs in formation-position order, maximizing
    /// each slot's OverallForPosition. Greedy: picks best available per slot.
    /// </summary>
    public static IReadOnlyList<int> OptimalLineup(
        IReadOnlyList<Player> squad, Formation formation)
    {
        var available = new HashSet<int>(squad.Select(p => p.Id));
        var playerMap = squad.ToDictionary(p => p.Id);
        int[] result = new int[formation.Positions.Count];

        // Sort slots by scarcity: GK first (fewest candidates), then others
        int[] slotOrder = Enumerable.Range(0, formation.Positions.Count)
            .OrderBy(i => formation.Positions[i] == Position.GK ? 0 : 1)
            .ThenBy(i => CountCandidates(squad, formation.Positions[i]))
            .ToArray();

        foreach (int slotIdx in slotOrder)
        {
            Position slotPos = formation.Positions[slotIdx];
            int bestId = -1;
            float bestOvr = float.MinValue;

            foreach (int pid in available)
            {
                Player p = playerMap[pid];
                float ovr = p.Attributes.OverallForPosition(slotPos);

                // Prefer players whose primary/secondary matches the slot
                if (p.PrimaryPosition == slotPos)
                    ovr += 5f;
                else if (p.SecondaryPosition == slotPos)
                    ovr += 2f;

                if (ovr > bestOvr)
                {
                    bestOvr = ovr;
                    bestId = pid;
                }
            }

            result[slotIdx] = bestId;
            available.Remove(bestId);
        }

        return result;
    }

    /// <summary>
    /// Calculates the average position-specific overall of an optimal lineup.
    /// </summary>
    public static float AverageOverall(IReadOnlyList<Player> squad, Formation formation)
    {
        IReadOnlyList<int> lineup = OptimalLineup(squad, formation);
        var playerMap = squad.ToDictionary(p => p.Id);

        float total = 0f;
        for (int i = 0; i < lineup.Count; i++)
        {
            if (playerMap.TryGetValue(lineup[i], out Player? player))
            {
                total += player.Attributes.OverallForPosition(formation.Positions[i]);
            }
        }

        return total / lineup.Count;
    }

    private static int CountCandidates(IReadOnlyList<Player> squad, Position pos)
    {
        return squad.Count(p => p.PrimaryPosition == pos || p.SecondaryPosition == pos);
    }
}

File: engine/simulation/formation-optimizer.ts
import { Position } from "@/engine/enums";
import type { Formation, Player } from "@/engine/models";
import { overallForPosition } from "@/engine/models";

/**
 * Greedy optimizer that assigns the best-fit player to each formation slot
 * based on position-specific overall ratings.
 */

/**
 * Returns 11 player IDs in formation-position order, maximizing
 * each slot's overallForPosition. Greedy: picks best available per slot.
 */
export function optimalLineup(
  squad: readonly Player[],
  formation: Formation,
): readonly number[] {
  const available = new Set(squad.map((p) => p.id));
  const playerMap = new Map(squad.map((p) => [p.id, p]));
  const result: number[] = new Array(formation.positions.length).fill(0);

  // Sort slots by scarcity: GK first (fewest candidates), then others
  const slotOrder = Array.from(
    { length: formation.positions.length },
    (_, i) => i,
  ).sort((a, b) => {
    const aIsGk = formation.positions[a] === Position.GK ? 0 : 1;
    const bIsGk = formation.positions[b] === Position.GK ? 0 : 1;
    if (aIsGk !== bIsGk) return aIsGk - bIsGk;
    return (
      countCandidates(squad, formation.positions[a]) -
      countCandidates(squad, formation.positions[b])
    );
  });

  for (const slotIdx of slotOrder) {
    const slotPos = formation.positions[slotIdx];
    let bestId = -1;
    let bestOvr = -Infinity;

    for (const pid of available) {
      const p = playerMap.get(pid)!;
      let ovr = overallForPosition(p.attributes, slotPos);

      // Prefer players whose primary/secondary matches the slot
      if (p.primaryPosition === slotPos) {
        ovr += 5;
      } else if (p.secondaryPosition === slotPos) {
        ovr += 2;
      }

      if (ovr > bestOvr) {
        bestOvr = ovr;
        bestId = pid;
      }
    }

    result[slotIdx] = bestId;
    available.delete(bestId);
  }

  return result;
}

/** Calculates the average position-specific overall of an optimal lineup. */
export function averageOverall(
  squad: readonly Player[],
  formation: Formation,
): number {
  const lineup = optimalLineup(squad, formation);
  const playerMap = new Map(squad.map((p) => [p.id, p]));

  let total = 0;
  for (let i = 0; i < lineup.length; i++) {
    const player = playerMap.get(lineup[i]);
    if (player != null) {
      total += overallForPosition(player.attributes, formation.positions[i]);
    }
  }

  return total / lineup.length;
}

function countCandidates(
  squad: readonly Player[],
  pos: Position,
): number {
  return squad.filter(
    (p) => p.primaryPosition === pos || p.secondaryPosition === pos,
  ).length;
}
